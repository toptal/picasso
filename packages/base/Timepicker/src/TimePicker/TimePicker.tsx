import React, { useEffect, useState } from 'react'
import type { BaseProps } from '@toptal/picasso-shared'
import InputMask from 'react-input-mask'
import { detect } from 'detect-browser'
import { Input } from '@toptal/picasso-input'
import { Time16 } from '@toptal/picasso-icons'
import type { InputProps } from '@toptal/picasso-input'
import type { Status } from '@toptal/picasso-outlined-input'
import { twMerge } from '@toptal/picasso-tailwind-merge'

export interface Props
  extends BaseProps,
    Omit<
      InputProps,
      | 'id'
      | 'value'
      | 'onSelect'
      | 'onChange'
      | 'type'
      | 'multiline'
      | 'rows'
      | 'defaultValue'
      | 'step'
      | 'icon'
      | 'iconPosition'
      | 'counter'
      | 'endAdornment'
      | 'startAdornment'
      | 'multilineResizable'
      | 'rowsMax'
      | 'limit'
      | 'placeholder'
      | 'status'
    > {
  /** Time value that will be selected in TimePicker */
  value?: string
  /** Indicate whether `TimePicker` is in `error` or `default` state */
  status?: Extract<Status, 'error' | 'default'>
  /** Called on input change */
  onChange?: (value: string) => void
}

const VALID_TIME_REGEX = new RegExp(/^([0-1][0-9]|2[0-3]):[0-5][0-9]$/)

export const TimePicker = (props: Props) => {
  const {
    onChange: externalOnChange,
    value: externalValue,
    width,
    className,
    status,
    highlight,
    ...rest
  } = props

  const [value, setValue] = useState(externalValue)

  useEffect(() => {
    // Set internal value based on the provided one if the later is correct
    if (externalValue && VALID_TIME_REGEX.test(externalValue)) {
      setValue(externalValue)
    }
  }, [externalValue])

  const onChange = (
    event: React.ChangeEvent<
      HTMLInputElement | HTMLTextAreaElement | HTMLSelectElement
    >
  ) => {
    const newValue = event.target.value

    setValue(newValue)

    if (newValue && VALID_TIME_REGEX.test(newValue)) {
      externalOnChange?.(newValue)
    } else {
      externalOnChange?.('')
    }
  }

  const browser = detect()
  const isSafari = browser?.name === 'safari'
  const startsWithTwo = value && value[0] === '2'

  const inputMask = [
    /[0-2]/,
    startsWithTwo ? /[0-3]/ : /[0-9]/,
    ':',
    /[0-5]/,
    /[0-9]/,
  ]

  const icon = (
    <Time16
      classes={{
        root: 'bg-white absolute right-[0.625rem] pointer-events-none m-0 select-none',
      }}
    />
  )

  const inputClassName = twMerge('cursor-default', className)

  const inputPropClassName = `-mr-[8px] [&::-webkit-calendar-picker-indicator]:absolute [&::-webkit-calendar-picker-indicator]:right-2
    [&::-webkit-calendar-picker-indicator]:cursor-pointer [&::-webkit-calendar-picker-indicator]:bg-none`

  if (isSafari) {
    return (
      <Input
        type='text'
        readOnly
        iconPosition='end'
        icon={icon}
        width={width}
        status={status}
        className={inputClassName}
        highlight={highlight}
        inputProps={{
          className: inputPropClassName,
          ...rest,
        }}
        startAdornment={
          <InputMask
            mask={inputMask}
            alwaysShowMask
            maskPlaceholder='-'
            value={value}
            onChange={onChange}
            className={'text-sm border-none p-0 m-0 outline-none'}
          />
        }
      />
    )
  }

  return (
    <Input
      type='time'
      value={value}
      className={inputClassName}
      onChange={onChange}
      iconPosition='end'
      highlight={highlight}
      icon={icon}
      width={width}
      status={status}
      inputProps={{
        className: inputPropClassName,
        step: 60, // 1 min
        ...rest,
      }}
    />
  )
}

TimePicker.defaultProps = {
  status: 'default',
}

TimePicker.displayName = 'TimePicker'

export default TimePicker
