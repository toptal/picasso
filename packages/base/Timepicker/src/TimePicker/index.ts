import type { OmitInternalProps } from '@toptal/picasso-shared'

import type { Props } from './TimePicker'

export { default as TimePicker } from './TimePicker'
export type TimePickerProps = OmitInternalProps<Props>
