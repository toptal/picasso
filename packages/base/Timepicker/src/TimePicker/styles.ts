import type { Theme } from '@material-ui/core/styles'
import { createStyles } from '@material-ui/core/styles'

export default ({ palette }: Theme) =>
  createStyles({
    root: {
      cursor: 'default',
    },
    icon: {
      background: palette.common.white,
      position: 'absolute',
      right: '0.625rem',
      pointerEvents: 'none',
      margin: 0,
      userSelect: 'none',
    },
    inputBase: {
      marginRight: '-8px', // override default margin for icon position

      // override styles for native 'clock' icon for type='time'
      '&::-webkit-calendar-picker-indicator': {
        outline: 'none',
        cursor: 'pointer',
      },
    },
    inputMask: {
      fontSize: '0.8125rem',
      border: 'none',
      padding: '0',
      margin: '0',
      outline: 'none',
    },
  })
