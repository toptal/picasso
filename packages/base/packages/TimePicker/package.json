{
  "name": "@toptal/picasso-time-picker",
  "version": "1.0.0",
  "main": "index.js",
  "dependencies": {
    "@toptal/picasso-input": "^1.0.0",
    "@toptal/picasso-icon": "^1.0.0",
    "@toptal/picasso-utils/use-deprecation-warnings": "^1.0.0",
    "@toptal/picasso-outlined-input": "^1.0.0"
  }
}