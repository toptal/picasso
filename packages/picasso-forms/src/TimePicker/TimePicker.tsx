import React from 'react'
import {
  TimePicker as PicassoTimePicker,
  TimePickerProps
} from '@toptal/picasso'

import { FieldProps } from '../Field'
import InputField from '../InputField'
import FieldLabel from '../FieldLabel'

export type FormTimePickerProps = Omit<TimePickerProps, 'onChange'> & {
  onChange?: TimePickerProps['onChange']
}
export type Props = FormTimePickerProps & FieldProps<TimePickerProps['value']>

export const TimePicker = (props: Props) => {
  const { label, titleCase, ...rest } = props

  return (
    <InputField<FormTimePickerProps>
      {...rest}
      label={
        label ? (
          <FieldLabel
            name={props.name}
            required={props.required}
            label={label}
            titleCase={titleCase}
          />
        ) : null
      }
    >
      {(inputProps: TimePickerProps) => {
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        const { enableReset, onResetClick, ...rest } = inputProps

        return <PicassoTimePicker {...rest} />
      }}
    </InputField>
  )
}

TimePicker.displayName = 'TimePicker'

export default TimePicker
