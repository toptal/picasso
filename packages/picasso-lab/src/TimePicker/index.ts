export { default } from './TimePicker'
export { Props as TimePickerProps } from './TimePicker'
