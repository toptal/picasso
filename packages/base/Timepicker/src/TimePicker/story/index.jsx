import { TimePicker } from '../TimePicker'
import PicassoBook from '~/.storybook/components/PicassoBook'

const page = PicassoBook.section('Forms').createPage(
  'TimePicker',
  `Time Picker component

  ${PicassoBook.createSourceLink(__filename)}
  `
)

page
  .createTabChapter('Props')
  .addComponentDocs({ component: TimePicker, name: 'TimePicker' })
page
  .createChapter()
  .addExample(
    'TimePicker/story/Default.example.tsx',
    {
      title: 'Default',
      takeScreenshot: false,
    },
    'base/Timepicker'
  )
  .addExample(
    'TimePicker/story/Status.example.tsx',
    {
      title: 'Status',
      takeScreenshot: false,
    },
    'base/Timepicker'
  )
