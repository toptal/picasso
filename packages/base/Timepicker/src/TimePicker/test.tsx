import React from 'react'
import { render, fireEvent } from '@toptal/picasso-test-utils'

import { TimePicker } from './TimePicker'

describe('TimePicker', () => {
  it('renders', () => {
    const time = '21:00'

    const { container } = render(
      <TimePicker value={time} onChange={() => {}} />
    )

    expect(container).toMatchSnapshot()
  })

  it('custom time rendering', () => {
    const time = '22:50'
    const nextTime = '18:30'
    const handleChange = jest.fn()

    const { getByDisplayValue } = render(
      <TimePicker value={time} onChange={handleChange} />
    )

    const input = getByDisplayValue(time)

    fireEvent.change(input, { target: { value: nextTime } })

    expect(handleChange).toHaveBeenCalledTimes(1)
  })

  describe('when invalid time is entered', () => {
    it('calls onChange with empty value', () => {
      const time = '09:00'
      const handleChange = jest.fn()
      const { getByDisplayValue } = render(
        <TimePicker value={time} onChange={handleChange} />
      )

      const input = getByDisplayValue(time)

      fireEvent.change(input, { target: { value: '12:--' } })
      expect(handleChange).toHaveBeenCalledTimes(1)
      expect(handleChange).toHaveBeenCalledWith('')

      const newTime = '12:12'

      fireEvent.change(input, { target: { value: newTime } })
      expect(handleChange).toHaveBeenCalledTimes(2)
      expect(handleChange).toHaveBeenCalledWith(newTime)
    })
  })
})
