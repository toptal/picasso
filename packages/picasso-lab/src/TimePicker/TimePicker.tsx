import React from 'react'
import { BaseProps } from '@toptal/picasso-shared'
import { Input, InputProps } from '@toptal/picasso'
import { Time16 } from '@toptal/picasso/Icon'
import { Theme, makeStyles } from '@material-ui/core/styles'
import InputMask from 'react-input-mask'
import { detect } from 'detect-browser'
import cx from 'classnames'

import styles from './styles'

const useStyles = makeStyles<Theme>(styles, {
  name: 'PicassoTimePicker'
})

export interface Props
  extends BaseProps,
    Omit<
      InputProps,
      | 'id'
      | 'value'
      | 'onSelect'
      | 'type'
      | 'multiline'
      | 'rows'
      | 'defaultValue'
      | 'step'
      | 'icon'
      | 'iconPosition'
      | 'counter'
      | 'endAdornment'
      | 'startAdornment'
      | 'enableReset'
      | 'onResetClick'
      | 'multilineResizable'
      | 'rowsMax'
      | 'limit'
      | 'placeholder'
    > {
  /** Time value that will be selected in TimePicker */
  value?: string
}

export const TimePicker = (props: Props) => {
  const { onChange, value, width, className, ...rest } = props
  const classes = useStyles()
  const browser = detect()
  const isSafari = browser?.name === 'safari'
  const startsWithTwo = value && value[0] === '2'

  const inputMask = [
    /[0-2]/,
    startsWithTwo ? /[0-3]/ : /[0-9]/,
    ':',
    /[0-5]/,
    /[0-9]/
  ]

  const icon = <Time16 classes={{ root: classes.icon }} />

  if (isSafari) {
    return (
      <Input
        type='text'
        readOnly
        iconPosition='end'
        icon={icon}
        width={width}
        className={cx(classes.root, className)}
        inputProps={{
          className: classes.inputBase,
          ...rest
        }}
        startAdornment={
          <InputMask
            mask={inputMask}
            alwaysShowMask
            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
            // @ts-ignore
            maskPlaceholder='-'
            value={value}
            onChange={onChange}
            className={classes.inputMask}
          />
        }
      />
    )
  }

  return (
    <Input
      type='time'
      value={value}
      className={cx(classes.root, className)}
      onChange={onChange}
      iconPosition='end'
      icon={icon}
      width={width}
      inputProps={{
        className: classes.inputBase,
        step: 60, // 1 min
        ...rest
      }}
    />
  )
}

TimePicker.displayName = 'TimePicker'

export default TimePicker
