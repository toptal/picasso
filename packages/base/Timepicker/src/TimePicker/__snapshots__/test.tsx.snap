// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimePicker renders 1`] = `
<div>
  <div
    class="Picasso-root"
  >
    <div
      class="base-Input base-Input base-Input relative inline-flex gap-y gap-x items-center rounded-sm [font-size:_unset] hover:[&_.resetButtonDirty]:visible p-2 h-8 w-[18.75rem] bg-white after:content-[""] after:inline-block after:absolute after:top-0 after:bottom-0 after:right-0 after:left-0 after:pointer-events after:border-solid after:rounded-sm after:border-gray [&:has(:focus)]:after:border-blue after:border [&:has(:focus)]:after:shadow-[0_0_0_3px] [&:has(:focus)]:after:shadow-blue hover:[&:not(:has(:focus))]:after:border-gray text-black cursor-default"
    >
      <input
        aria-invalid="false"
        autocomplete="none"
        class="base-Input bg-transparent border-none box-border cursor-[inherit] h-full outline-none p-0 peer resize-none w-full text-md leading-4 text-black [&::placeholder]:text-gray [&::placeholder]:opacity-100 -mr [&::-webkit-calendar-picker [&::-webkit-calendar-picker-indicator]:right [&::-webkit-calendar-picker-indicator]:cursor [&::-webkit-calendar-picker-indicator]:bg"
        step="60"
        type="time"
        value="21:00"
      />
      <div
        class="text-graphite h-auto flex items-center whitespace-nowrap max-h justify-end ml-auto flex-grow flex-shrink basis-auto pointer-events"
      >
        <svg
          class="PicassoSvgTime16-root bg-white absolute right-[0.625rem] pointer-events m-0 select-none grow shrink basis-0"
          style="min-width: 16px; min-height: 16px;"
          viewBox="0 0 16 16"
        >
          <path
            d="M7.5 15a7.5 7.5 0 1 1 0-15 7.5 7.5 0 0 1 0 15Zm0-1a6.5 6.5 0 1 0 0-13 6.5 6.5 0 0 0 0 13ZM7 8H4V7h3V2h1v6H7Z"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;
